### PR TITLE
GDAL/GRASS driver: simplified to GRASS GIS 7+ only

### PR DESCRIPTION
**Pull request by neteler**
## What does this PR do?

This PR is a code simplification which removes the support for the old GRASS GIS 5 and 6 version which are out of maintenance for many years.

Open questions:

- shall `gdal/frmts/grass/grass57dataset.cpp` be renamed to `gdal/frmts/grass/grassdataset.cpp`?
- shall the driver name remain `GRASS57 driver`?

```
gdal/frmts/grass/grass57dataset.cpp
1021:    if( !GDAL_CHECK_VERSION( "GDAL/GRASS57 driver" ) )
```

It is ok for me to get this PR rejected in case that support for GRASS GIS 5 and 6 must remain.

Potential reviewer: @metzm 

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [x] Add test case(s)
 - [x] Add documentation
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

**Comment by rouault** [comment]
> * shall `gdal/frmts/grass/grass57dataset.cpp` be renamed to `gdal/frmts/grass/grassdataset.cpp`?

yes, please ! I've alway found the current naming confusing

> * shall the driver name remain `GRASS57 driver`?

I'd suggest to rename it to GRASS

> in case that support for GRASS GIS 5 and 6 must remain.

Personnaly I don't see the point in supporting those outdated versions.

doc/source/drivers/raster/grass.rst would perhaps need some updates toward its bottom

File: gdal/frmts/grass/grass57dataset.cpp [hunk 1]
```diff
@@ -454,23 +416,23 @@ CPLErr GRASSRasterBand::ResetReading ( struct Cell_head *sNewWindow )
          sNewWindow->ew_res != sOpenWindow.ew_res || sNewWindow->ns_res != sOpenWindow.ns_res ||
          sNewWindow->rows   != sOpenWindow.rows   || sNewWindow->cols   != sOpenWindow.cols )
     {
-        SetWindow ( sNewWindow );
-        G_copy((void *) &sOpenWindow, (void *) sNewWindow, sizeof(struct Cell_head));
+        Rast_set_window ( sNewWindow );
```
Comment (rouault) [inline gdal/frmts/grass/grass57dataset.cpp:419]: this change is wrong. The call to the C++ SetWindow() function is intended.

File: gdal/frmts/grass/grass57dataset.cpp [hunk 2]
```diff
@@ -454,23 +416,23 @@ CPLErr GRASSRasterBand::ResetReading ( struct Cell_head *sNewWindow )
...
+        Rast_get_window ( &sCurrentWindow );
 
         if ( sNewWindow->north  != sCurrentWindow.north  || sNewWindow->south  != sCurrentWindow.south ||
              sNewWindow->east   != sCurrentWindow.east   || sNewWindow->west   != sCurrentWindow.west ||
              sNewWindow->ew_res != sCurrentWindow.ew_res || sNewWindow->ns_res != sCurrentWindow.ns_res ||
              sNewWindow->rows   != sCurrentWindow.rows   || sNewWindow->cols   != sCurrentWindow.cols
              )
         {
-            SetWindow ( sNewWindow );
+            Rast_set_window ( sNewWindow );
```
Comment (rouault) [inline gdal/frmts/grass/grass57dataset.cpp:435]: this change is wrong. The call to the C++ SetWindow() function is intended.

**Review by rouault — CHANGES_REQUESTED** [review]

**Comment by rouault** [comment]
Have you tried running autotest/gdrivers/grass.py locally ? There isn't much in it. I don't currently have a GRASS setup. And our CI doesn't build with the GRASS driver

File: gdal/frmts/grass/GNUmakefile [hunk 3]
```diff
@@ -7,7 +7,7 @@ include ../../GDALmake.opt
 ifeq ($(GRASS_SETTING),libgrass)
 OBJ	=	grassdataset.o
 else
-OBJ	=	grass57dataset.o
+OBJ	=	grass.o
```
Comment (rouault) [inline gdal/frmts/grass/GNUmakefile:10]: the old grassdataset.cpp (was this for GRASS 4 ?) should be removed too, as well as traces of it like 2 lines above

File: gdal/frmts/grass/grass57dataset.cpp (hunk 1 above)
Comment (neteler) [reply gdal/frmts/grass/grass57dataset.cpp:419]: Thanks, hope I got it right in the updated version (now: grass.cpp)

File: gdal/frmts/grass/grass57dataset.cpp (hunk 2 above)
Comment (neteler) [reply gdal/frmts/grass/grass57dataset.cpp:435]: Thanks, hope I got it right in the updated version (now: grass.cpp)

File: gdal/doc/source/drivers/raster/grass.rst [hunk 4]
```diff
@@ -69,15 +69,15 @@ Notes on driver variations
 --------------------------
 
 For GRASS 5.7 Radim Blazek has moved the driver to using the GRASS
-shared libraries directly instead of using libgrass. Currently (GDAL
-1.2.2 and later) both version of the driver are available and can be
-configured using ``--with-libgrass`` for the libgrass variant or
-``--with-grass=<dir>`` for the new GRASS 5.7+ library based version. The
-GRASS 5.7+ driver version is currently not supporting coordinate system
-access, though it is hoped that will be corrected at some point.
+shared libraries directly instead of using libgrass. Currently both
```
Comment (rouault) [inline gdal/doc/source/drivers/raster/grass.rst:72]: this needs to be revised

File: gdal/frmts/grass/GNUmakefile (hunk 3 above)
Comment (neteler) [reply gdal/frmts/grass/GNUmakefile:5]: > Have you tried running autotest/gdrivers/grass.py locally ? There isn't much in it. I don't currently have a GRASS setup. And our CI doesn't build with the GRASS driver

Yes, it runs fine:

```
23:37 $ pytest gdrivers/grass.py
Test session starts (platform: linux, Python 3.8.5, pytest 4.6.11, pytest-sugar 0.9.4)
rootdir: /home/mneteler/software/gdal/autotest, inifile: pytest.ini
plugins: sugar-0.9.4, env-0.6.2
collecting ... 
 gdrivers/grass.py ✓✓                                                                                                                                        100% 

Results (0.29s):
       2 passed

```

File: gdal/doc/source/drivers/raster/grass.rst (hunk 4 above)
Comment (neteler) [reply gdal/doc/source/drivers/raster/grass.rst:72]: I am not sure how to simplify `gdal/configure.ac`.

Comment (rouault) [reply gdal/doc/source/drivers/raster/grass.rst:72]: I see @landam was the last one to too the GRASS part in configure.ac. My understanding of it + your changes is that this should still work unmodified against GRASS 7.x. Simplifying configure.ac would be indeed nice, but could be a work for later I guess

Comment (neteler) [reply gdal/doc/source/drivers/raster/grass.rst:72]: That's of course fine for me.

Comment (rouault) [reply gdal/doc/source/drivers/raster/grass.rst:72]: my point was about just the .rst. The mentions for "new GRASS 5.7+" could be simplified to just mentioning GRASS 7

File: gdal/frmts/grass/grass57dataset.cpp (hunk 2 above)
Comment (rouault) [reply gdal/frmts/grass/grass57dataset.cpp:408]: You should check with @metzm who added SetWindow() very recently per 28a38e366efaa59aaad9cb074d2dbf8358aa9c13

Comment (neteler) [reply gdal/frmts/grass/grass57dataset.cpp]: Now I got it thanks to this reference. Fixed and pytest passes again.

File: gdal/frmts/grass/grass57dataset.cpp (hunk 1 above)
Comment (neteler) [reply gdal/frmts/grass/grass57dataset.cpp]: Now I got it thanks to the reference 28a38e3. Fixed and pytest passes again.

File: gdal/doc/source/drivers/raster/grass.rst (hunk 4 above)
Comment (neteler) [reply gdal/doc/source/drivers/raster/grass.rst]: Updated accordingly.

**Comment by neteler** [comment]
Mhh,, not sure why this single test failed?

```
ogr/ogr_vfk.py::test_ogr_vfk_9 PASSED                                    [ 25%]
ogr/ogr_vfk.py::test_ogr_vfk_10 PASSED                                   [ 25%]
ogr/ogr_vfk.py::test_ogr_vfk_11 
Command exited with code 
```

**Comment by rouault** [comment]
> Mhh,, not sure why this single test failed?

Probably a flake. I've restarted that build

**Comment by rouault** [comment]
@neteler Looks good to me. Are you done with this PR ?

**Comment by neteler** [comment]
@rouault thanks for your guidance here! From my side I am done but

- I'd appreciate the review by @metzm - I'll contact him
- (pls reset the "Changes requested" flag if done for you)

**Review by rouault — APPROVED** [review]

File: gdal/doc/source/drivers/raster/grass.rst [hunk 5]
```diff
@@ -37,13 +37,13 @@ GRASS raster maps/imagery groups can be selected in several ways.
...
+      gdalinfo /data/grassdb/imagery/raw/group/testmff/REF
+      gdalinfo /data/grassdb/imagery/raw/group/testmff
 
-#. If there is a correct ``.grassrc5`` (GRASS 5), ``.grassrc6`` (GRASS
-   6) ``.grass7/rc`` (GRASS 7) setup file in the users home directory
-   then raster maps or imagery groups may be opened just by the cell
-   name. This only works for raster maps or imagery groups in the
+#. If there is a correct ``.grassrc7/rc`` (GRASS 7) setup file in the
+   user's home directory then raster maps or imagery groups may be opened
+   just by the cell name.
```
Comment (metzm) [inline gdal/doc/source/drivers/raster/grass.rst:45]:
```suggestion
   just by the cell or group name.
```

**Review by metzm — APPROVED** [review]
These changes are simplifying the code quite a bit, further maintenance of the GDAL-GRASS driver will be easier.

**Comment by neteler** [comment]
@rouault it may be merged now.

**Pull request merged**